sdks/java/harness/src/main/java/org/apache/beam/fn/harness/state/StateFetchingIterators.java

```diff
@@ -195,8 +245,17 @@ public ByteString firstPage() {
       this.continuationToken = stateRequestForFirstChunk.getGet().getContinuationToken();
     }
 
-    private void prefetch() {
-      if (prefetchedResponse == null && currentState == State.READ_REQUIRED) {
+    @Override
+    public boolean isReady() {
+      if (prefetchedResponse == null) {
+        return currentState != State.READ_REQUIRED;
+      }
+      return prefetchedResponse.isDone();
+    }
+
+    @Override
+    public void prefetch() {
+      if (!isReady()) {
         prefetchedResponse = new CompletableFuture<>();
```

Comment:
Is there a risk of overriding an ongoing prefetch? Do we need to check that prefetchedResponse is null?

Reply:
Yes.

Reply:
Yes